ls1intum/Artemis
# `General`: Allow admins to filter user management table

### Checklist
#### General


#### Server
#### Client

### Motivation and Context


Prepare #5061 and provides a filterable user management view

### Description

In order to still provide the possibilities of pages, i integrated the filter options into the page and filter the users on the database level. This allows us to still use the benefits of pages.

We provide several filters:

- **Role**: Filter the users according to their permissions/role _(stored in local storage)_
- **Origin**: Filters the users based on their administrative location (internal or external) _(stored in local storage)_
- **Status**: Filter the users based on their account status (activated or deactivated) _(stored in local storage)_
- **Course**: Filters the users according to their course affiliation

The first time we visit the page, we have to load all the courses (which takes some time). After that, the filtering should be really fast.

### Steps for Testing

Prerequisites:
- 1 Admin

1. Log in to Artemis
2. Visit the user management page
3. Add/Remove users, apply filters and so on
4. Verify that everything works

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| user-management.component.ts | 73.68% | 83.81% |
| UserResource | 89% | 75% |

### Screenshots



![image](https://user-images.githubusercontent.com/60005702/172572876-f65b39db-e1ef-467e-a97c-0483c17f29dc.png)
![image](https://user-images.githubusercontent.com/60005702/172572943-99fd3cf6-c16e-48dd-acb9-473d5745d14c.png)
![image](https://user-images.githubusercontent.com/60005702/172572991-11406c6b-b8c4-4955-ae94-4cb23f372294.png)
